servers/rendering/renderer_rd/shaders/effects/tonemap.glsl

```diff
@@ -244,7 +244,8 @@ vec3 tonemap_aces(vec3 color, float white) {
 }
 
 vec3 tonemap_reinhard(vec3 color, float white) {
-	return (white * color + color) / (color * white + white);
+	float white_squared = white * white;
+	return (white_squared * color + color * color) / (color * white_squared + white_squared);
```

Comment:
```suggestion
	return (white_squared * color + color) / (color * white_squared + white_squared);
```

Looks like an extra `color` slipped in here.

Reply:
The extra color is part of the fix. It is needed to conform to the formula given in Reinhard's paper. That it was missing in Godot's implementation is the main issue.

Reply:
Fixed the implementation for the compatibility renderer as well - thanks, I don't know how I missed that!

Reply:
I haven't looked at the paper, but from your issue report, you say that Reinhard's formula is:
```
color * (1 + 1/(white * white)) / (color + 1)
```
While we use:
```
(white * color + color) / (color * white + white);
```
If you divide the entire thing by `white` you get:
```
(color + (color/white)) / (color + 1)
```
And if you factor out `color` from the top half you get Reinhard's original formula (with white being broken of course):
```
color * (1 + 1/(white)) / (color + 1)
```

This new version you have added turns into:
```
color * (1 + color / (white * white)) / (color + 1)
```

Reply:
Ah, I'm sorry, I made a mistake in my issue report! The formula (4) in Reinhard's paper is
```
color * (1 + color/(white * white)) / (color + 1)
```
![Screenshot_20240618_235824](https://github.com/godotengine/godot/assets/155747722/d2ad9cfc-29f6-48ee-8866-1edd7e0d02e4)
L(x,y) is our `color`, and L_white is `white`. So my fix is correct, I just made a typo in the issue, gonna fix that right away.

Reply:
PR as it stands [appears to be correct](https://www.desmos.com/calculator/2wcdxmi8dq).

I presume that it's better performance to be written like this, otherwise matching the original formula's format would be a touch easier to read if someone is comparing it to the paper in the future.

Reply:
Ah, that makes sense. So this is a bigger change than we thought. 

> I presume that it's better performance to be written like this, otherwise matching the original formula's format would be a touch easier to read if someone is comparing it to the paper in the future.

Ya, performance is better the way it was originally written. 

**OLD**
This is 3 instructions (1 divide, 2 FMAs)
```
(white * color + color) / (color * white + white);
```

This is 5 instructions (2 divides, 2 additions, 1 multiply):
```
color * (1 + 1/(white)) / (color + 1)
```

**New**
This is 5 instructions (2 multiply, 2 FMA, 1 divide)
```
float white_squared = white * white;
(white_squared * color + (color * color)) / (color * white_squared + white_squared);
```

This is 6 instructions (2 multiply, 2 addition, 2 divide)
```
color * (1 + color / (white * white)) / (color + 1)
```

Reply:
I added comments to the code showing the non-simplified formula and a link to the paper, so in the future people can more easily check the source.

Reply:
```suggestion
	float white_squared_color = white_squared * color;
	return (white_squared_color  + color * color) / (white_squared_color  + white_squared);
```

`white_squared * color` and `color * white_squared` are identical? - I suggest pushing it to float constant and use it in both places.

Reply:
```suggestion
	return fma(white_squared, color, color * color) / fma(color, white_squared, white_squared);
```

If the suggestion above is not desired, consider to use built-in `fma` function.

Reply:
Good catch, thank you!